File: expressoes/src/main/java/com/tiduswr/Calculador.java
package com.tiduswr;

import com.tiduswr.ExpressoesParser.ExpressaoContext;
import com.tiduswr.ExpressoesParser.FatorContext;
import com.tiduswr.ExpressoesParser.ProgramaContext;
import com.tiduswr.ExpressoesParser.TermoContext;

public class Calculador extends ExpressoesBaseVisitor<Double>{
    
    @Override
    public Double visitPrograma(ProgramaContext ctx) {
        return visitExpressao(ctx.expressao());
    }

    @Override
    public Double visitExpressao(ExpressaoContext ctx) {
        var valor = visitTermo(ctx.t1);
        
        for(int i = 0; i < ctx.termos.size(); i++){
            var op1 = ctx.op1(i);
            var termo2 = ctx.termos.get(i);

            if(op1.getText().equals("+")){
                valor += visitTermo(termo2);
            }else{
                valor -= visitTermo(termo2);
            }
        }

        return valor;
    }

    @Override
    public Double visitTermo(TermoContext ctx) {
        var valor = visitFator(ctx.f1);

        for(int i = 0; i < ctx.fatores.size(); i++){
            var op2 = ctx.op2(i);
            var fator2 = ctx.fatores.get(i);

            if(op2.getText().equals("*")){
                valor *= visitFator(fator2);
            }else{
                valor /= visitFator(fator2);
            }
        }

        return valor;
    }

    @Override
    public Double visitFator(FatorContext ctx) {
        if(ctx.NUM() != null){
            return Double.parseDouble(ctx.NUM().getText());
        }else{
            return visitExpressao(ctx.expressao());
        }
    }

}

File: antlr-python-example/Calculador.py
from build.ExpressoesVisitor import ExpressoesVisitor
from build.ExpressoesParser import ExpressoesParser as ExprParser

class Calculador(ExpressoesVisitor):
    def visitPrograma(self, ctx:ExprParser.ProgramaContext):
        return self.visitExpressao(ctx.expressao())

    def visitExpressao(self, ctx:ExprParser.ExpressaoContext):
        valor = self.visitTermo(ctx.t1)

        for i in range(len(ctx.termos)):
            op1 = ctx.op1(i)
            termo2 = ctx.termos[i]

            if op1.getText() == "+":
                valor += self.visitTermo(termo2)
            else:
                valor -= self.visitTermo(termo2)
        
        return valor

    def visitTermo(self, ctx:ExprParser.TermoContext):
        valor = self.visitFator(ctx.f1)

        for i in range(len(ctx.fatores)):
            op1 = ctx.op2(i)
            fator2 = ctx.fatores[i]

            if(op1.getText() == "*"):
                valor *= self.visitFator(fator2)
            else:
                valor /= self.visitFator(fator2)
        
        return valor
    
    def visitFator(self, ctx:ExprParser.FatorContext):
        if ctx.NUM():
            return float(ctx.NUM().getText())
        else:
            return self.visitExpressao(ctx.expressao())